### PR TITLE
DRL-36-created custom metadata and changes

### PR DESCRIPTION
**Pull request by Wafashk**
created custom metadata and changes in test class and TriggerHelper

### CODE REVIEW

File: force-app/main/default/classes/DRL_LeadTriggerHelper.cls [hunk 1]
```diff
@@ -246,5 +246,59 @@ public without sharing class DRL_LeadTriggerHelper {
...
+                               where Id In: set_prospectId]){
+            set_OldParentStatus.add(objparentLead.Status);
+                                   
+            
+        }
+        
+        for(Prospect_Lead_Status_Mapping__mdt objprospectLeadStatus :[Select Lead_Status__c,
+                                                                      Weightage__c,
+                                                                      Prospect_Status__c
+                                                                      from Prospect_Lead_Status_Mapping__mdt
```
Comment (csankit0108) [inline force-app/main/default/classes/DRL_LeadTriggerHelper.cls:270]:
```suggestion
                                                                      FROM Prospect_Lead_Status_Mapping__mdt
```

File: force-app/main/default/classes/DRL_LeadTriggerHelper.cls [hunk 2]
```diff
@@ -246,5 +246,59 @@ public without sharing class DRL_LeadTriggerHelper {
...
+                                                                      where Lead_Status__c In:set_OldParentStatus]){
```
Comment (csankit0108) [inline force-app/main/default/classes/DRL_LeadTriggerHelper.cls:271]:
```suggestion
                                                                      WHERE Lead_Status__c IN :set_OldParentStatus]) {
```

File: force-app/main/default/classes/DRL_LeadTriggerHelper.cls [hunk 3]
```diff
@@ -246,5 +246,59 @@ public without sharing class DRL_LeadTriggerHelper {
...
+                                                                       map_getOldWeightage.put(objprospectLeadStatus.Lead_Status__c,objprospectLeadStatus.Weightage__c);
```
Comment (csankit0108) [inline force-app/main/default/classes/DRL_LeadTriggerHelper.cls:272]:
```suggestion
                                                                       map_getOldWeightage.put(objprospectLeadStatus.Lead_Status__c, objprospectLeadStatus.Weightage__c);
```

File: force-app/main/default/classes/DRL_LeadTriggerHelper.cls [hunk 4]
```diff
@@ -246,5 +246,59 @@ public without sharing class DRL_LeadTriggerHelper {
...
+                                                                       dbloldWeightage = objprospectLeadStatus.Weightage__c;
+                                                                       System.debug('old parent status' + dbloldWeightage );
```
Comment (csankit0108) [inline force-app/main/default/classes/DRL_LeadTriggerHelper.cls:274]: Remove system debugs

File: force-app/main/default/classes/DRL_LeadTriggerHelper.cls [hunk 5]
```diff
@@ -246,5 +246,59 @@ public without sharing class DRL_LeadTriggerHelper {
...
+                                                                   }
+                                                                 
+        
+        for(Prospect_Lead_Status_Mapping__mdt objprospectLeadStatus :[Select Lead_Status__c,
```
Comment (csankit0108) [inline force-app/main/default/classes/DRL_LeadTriggerHelper.cls:278]: Can we combine this FOR loop with the above one

File: force-app/main/default/classes/DRL_LeadTriggerHelper.cls [hunk 6]
```diff
@@ -246,5 +246,59 @@ public without sharing class DRL_LeadTriggerHelper {
...
+        for(Lead objparentLead : [select Id,
+                               Status 
+                               from Lead 
+                               where Id In: set_prospectId]){
+            set_OldParentStatus.add(objparentLead.Status);
+                                   
+            
+        }
+        
+        for(Prospect_Lead_Status_Mapping__mdt objprospectLeadStatus :[Select Lead_Status__c,
```
Comment (csankit0108) [inline force-app/main/default/classes/DRL_LeadTriggerHelper.cls:267]: Update the custom metadata api names correctly in the code

File: force-app/main/default/classes/DRL_LeadTriggerHelper.cls [hunk 7]
```diff
@@ -246,5 +246,59 @@ public without sharing class DRL_LeadTriggerHelper {
...
+                                                                   Prospect_Status__c 
+                                                                   from Prospect_Lead_Status_Mapping__mdt
+                                                                   where Lead_Status__c In: set_newLeadStatus 
+                                                                   ]){
+                                                                      strnewParentStatus = objprospectLeadStatus.Prospect_Status__c;
+                                                                      dblnewWeightage = objprospectLeadStatus.Weightage__c;
+                                                                  }
+        
+        
+        for(Lead objparentLead : [select Id,
```
Comment (csankit0108) [inline force-app/main/default/classes/DRL_LeadTriggerHelper.cls:289]:
```suggestion
        for (Lead objLead : [SELECT Id,
```

File: force-app/main/default/classes/DRL_LeadTriggerHelper.cls [hunk 8]
```diff
@@ -246,5 +246,59 @@ public without sharing class DRL_LeadTriggerHelper {
...
+                               Status
+                               from Lead
```
Comment (csankit0108) [inline force-app/main/default/classes/DRL_LeadTriggerHelper.cls:291]:
```suggestion
                               FROM Lead
```

File: force-app/main/default/classes/DRL_LeadTriggerHelper.cls [hunk 9]
```diff
@@ -246,5 +246,59 @@ public without sharing class DRL_LeadTriggerHelper {
...
+                               where Id In: set_prospectId]){
```
Comment (csankit0108) [inline force-app/main/default/classes/DRL_LeadTriggerHelper.cls:292]:
```suggestion
                               WHERE Id In :set_prospectId]) {
```

File: force-app/main/default/classes/DRL_LeadTriggerHelper.cls [hunk 10]
```diff
@@ -246,5 +246,59 @@ public without sharing class DRL_LeadTriggerHelper {
...
+            if(dblnewWeightage> dbloldWeightage  && strnewParentStatus != objparentLead.Status)
```
Comment (csankit0108) [inline force-app/main/default/classes/DRL_LeadTriggerHelper.cls:293]:
```suggestion
            if (dblnewWeightage > dbloldWeightage  && strnewParentStatus != objparentLead.Status)
```

File: force-app/main/default/classes/DRL_LeadTriggerHelper.cls [hunk 11]
```diff
@@ -246,5 +246,59 @@ public without sharing class DRL_LeadTriggerHelper {
...
+            {
+
```
Comment (csankit0108) [inline force-app/main/default/classes/DRL_LeadTriggerHelper.cls:295]: remove empty line

File: force-app/main/default/classes/DRL_LeadTriggerHelper.cls [hunk 12]
```diff
@@ -246,5 +246,59 @@ public without sharing class DRL_LeadTriggerHelper {
...
+        for(Lead objparentLead : [select Id,
+                               Status
+                               from Lead
+                               where Id In: set_prospectId]){
+            if(dblnewWeightage> dbloldWeightage  && strnewParentStatus != objparentLead.Status)
+            {
+                
+                objparentLead.Status = strnewParentStatus;
+                list_updateParentLead.add(objparentLead);
+            }
```
Comment (csankit0108) [inline force-app/main/default/classes/DRL_LeadTriggerHelper.cls:298]: add new line

File: force-app/main/default/classes/DRL_LeadTriggerHelper.cls [hunk 13]
```diff
@@ -246,5 +246,59 @@ public without sharing class DRL_LeadTriggerHelper {
...
+            if(list_updateParentLead.size() > 0){
```
Comment (csankit0108) [inline force-app/main/default/classes/DRL_LeadTriggerHelper.cls:299]:
```suggestion
            if (!list_updateParentLead.isEmpty()) {
```

File: force-app/main/default/classes/DRL_LeadTriggerHelper.cls [hunk 14]
```diff
@@ -246,5 +246,59 @@ public without sharing class DRL_LeadTriggerHelper {
...
+                update list_updateParentLead;
```
Comment (csankit0108) [inline force-app/main/default/classes/DRL_LeadTriggerHelper.cls:300]: Use Database.update and add errors to exception logs

File: force-app/main/default/classes/DRL_LeadTriggerHelperTest.cls [hunk 15]
```diff
@@ -34,16 +35,32 @@ public class DRL_LeadTriggerHelperTest {
 
         Map<String, Id> map_RecordTypeIdByDeveloperName = DRLUtil.getRecordTypeDeveloperNameIdMap('Lead');
         
-        Lead objLead = new DRL_TestUtility.LeadBuilder()
+        // Lead objLead = new DRL_TestUtility.LeadBuilder()
```
Comment (csankit0108) [inline force-app/main/default/classes/DRL_LeadTriggerHelperTest.cls:38]: Remove commented lines

File: force-app/main/default/classes/DRL_LeadTriggerHelperTest.cls [hunk 16]
```diff
@@ -126,4 +143,32 @@ public class DRL_LeadTriggerHelperTest {
         blnThrowException = false;
         Test.stopTest();
     }
+    /**
+    * @ author       :  Wafa Shaikh
+    * @ description  :  This method is used to test the Prospect Status When the  child Lead is updated 
+    **/
+    @isTest
+    private static void TestProspectStatus(){
```
Comment (csankit0108) [inline force-app/main/default/classes/DRL_LeadTriggerHelperTest.cls:151]:
```suggestion
    private static void testProspectStatus() {
```

File: force-app/main/default/classes/DRL_LeadTriggerHelperTest.cls [hunk 17]
```diff
@@ -126,4 +143,32 @@ public class DRL_LeadTriggerHelperTest {
...
+        Map<String, Id> map_RecordTypeIdByDeveloperName = DRLUtil.getRecordTypeDeveloperNameIdMap('Lead');
+ 
+      Lead objLead=[Select Id,
```
Comment (csankit0108) [inline force-app/main/default/classes/DRL_LeadTriggerHelperTest.cls:154]:
```suggestion
      Lead objLead = [SELECT Id,
```

File: force-app/main/default/classes/DRL_LeadTriggerHelperTest.cls [hunk 18]
```diff
@@ -126,4 +143,32 @@ public class DRL_LeadTriggerHelperTest {
...
+                      Email,
+                      Status,	
+                      DRL_Prospect__c 
+                      From Lead
```
Comment (csankit0108) [inline force-app/main/default/classes/DRL_LeadTriggerHelperTest.cls:158]:
```suggestion
                      FROM Lead
```

File: force-app/main/default/classes/DRL_LeadTriggerHelperTest.cls [hunk 19]
```diff
@@ -126,4 +143,32 @@ public class DRL_LeadTriggerHelperTest {
...
+                      Where Email='testuser5@yopmail.com'  and
```
Comment (csankit0108) [inline force-app/main/default/classes/DRL_LeadTriggerHelperTest.cls:159]:
```suggestion
                      WHERE Email ='testuser5@yopmail.com' AND
```

File: force-app/main/default/classes/DRL_LeadTriggerHelperTest.cls [hunk 20]
```diff
@@ -126,4 +143,32 @@ public class DRL_LeadTriggerHelperTest {
...
+        Map<String, Id> map_RecordTypeIdByDeveloperName = DRLUtil.getRecordTypeDeveloperNameIdMap('Lead');
+ 
+      Lead objLead=[Select Id,
+                      Email,
+                      Status,	
+                      DRL_Prospect__c 
+                      From Lead
+                      Where Email='testuser5@yopmail.com'  and 
+                      RecordTypeId = :map_RecordTypeIdByDeveloperName.get('DRL_Prospect') 
+
```
Comment (csankit0108) [inline force-app/main/default/classes/DRL_LeadTriggerHelperTest.cls:161]: remove empty lines

File: force-app/main/default/classes/DRL_LeadTriggerHelperTest.cls [hunk 21]
```diff
@@ -126,4 +143,32 @@ public class DRL_LeadTriggerHelperTest {
...
+                      Email,
+                      Status,	
+                      DRL_Prospect__c 
+                      From Lead
+                      Where Email='testuser5@yopmail.com'  and 
+                      RecordTypeId = :map_RecordTypeIdByDeveloperName.get('DRL_Prospect') 
+                       
+                   
+                     ];
+     system.debug('record fetched');
```
Comment (csankit0108) [inline force-app/main/default/classes/DRL_LeadTriggerHelperTest.cls:164]: remove system debugs

File: force-app/main/default/classes/DRL_LeadTriggerHelperTest.cls [hunk 22]
```diff
@@ -126,4 +143,32 @@ public class DRL_LeadTriggerHelperTest {
...
+        
+     Test.startTest();
+        
+        objLead.Status='Unqualified';
```
Comment (csankit0108) [inline force-app/main/default/classes/DRL_LeadTriggerHelperTest.cls:168]:
```suggestion
        objLead.Status = 'Unqualified';
```

File: force-app/main/default/classes/DRL_LeadTriggerHelperTest.cls [hunk 23]
```diff
@@ -126,4 +143,32 @@ public class DRL_LeadTriggerHelperTest {
...
+        Update objLead;
```
Comment (csankit0108) [inline force-app/main/default/classes/DRL_LeadTriggerHelperTest.cls:169]:
```suggestion
        update objLead;        
```

File: force-app/main/default/classes/DRL_LeadTriggerHelperTest.cls [hunk 24]
```diff
@@ -126,4 +143,32 @@ public class DRL_LeadTriggerHelperTest {
...
+                   
+                     ];
+     system.debug('record fetched');
+        
+     Test.startTest();
+        
+        objLead.Status='Unqualified';
+        Update objLead;        
+        Test.stopTest();
+
```
Comment (csankit0108) [inline force-app/main/default/classes/DRL_LeadTriggerHelperTest.cls:171]: remove empty lines

File: force-app/main/default/classes/DRL_TestUtility.cls [hunk 25]
```diff
@@ -69,7 +75,23 @@ public class DRL_TestUtility {
             this.strUTMSource = strValue;
             return this;
         }
-    
+        public LeadBuilder setProspectId(Id idValue) {
+            this.idProspectId = idValue;
+            return this;
+        }
```
Comment (csankit0108) [inline force-app/main/default/classes/DRL_TestUtility.cls:81]: add new line after every method ends

File: force-app/main/default/classes/DRL_TestUtility.cls [hunk 26]
```diff
@@ -69,7 +75,23 @@ public class DRL_TestUtility {
...
+        public LeadBuilder setStatus (String strValue) {
```
Comment (csankit0108) [inline force-app/main/default/classes/DRL_TestUtility.cls:82]:
```suggestion
        public LeadBuilder setStatus(String strValue) {
```

File: force-app/main/default/classes/DRL_TestUtility.cls [hunk 27]
```diff
@@ -69,7 +75,23 @@ public class DRL_TestUtility {
...
+            this.strStatus = strValue;
+            return this;
+        }
+        public LeadBuilder setCountry (String strValue){
```
Comment (csankit0108) [inline force-app/main/default/classes/DRL_TestUtility.cls:86]:
```suggestion
        public LeadBuilder setCountry(String strValue) {
```

File: force-app/main/default/classes/DRL_TestUtility.cls [hunk 28]
```diff
@@ -69,7 +75,23 @@ public class DRL_TestUtility {
...
+            this.strCountry=strValue;
+            return this;
+        }
+        
+         public LeadBuilder setReasonForDecline (String strValue){
```
Comment (csankit0108) [inline force-app/main/default/classes/DRL_TestUtility.cls:91]:
```suggestion
         public LeadBuilder setReasonForDecline(String strValue){
```

File: force-app/main/default/classes/DRL_TestUtility.cls [hunk 29]
```diff
@@ -79,7 +101,10 @@ public class DRL_TestUtility {
             objLead.DRL_UTMCampaign__c = this.strUTMCampaign;
             objlead.DRL_UTMSource__c = this.strUTMSource;
             objLead.RecordTypeId = this.idRecordTypeId;
-            
+            objLead.DRL_Prospect__c=this.idProspectId;
```
Comment (csankit0108) [inline force-app/main/default/classes/DRL_TestUtility.cls:104]:
```suggestion
            objLead.DRL_Prospect__c = this.idProspectId;
```

File: force-app/main/default/classes/DRL_TestUtility.cls [hunk 30]
```diff
@@ -79,7 +101,10 @@ public class DRL_TestUtility {
...
+            objLead.Status=this.strStatus;
```
Comment (csankit0108) [inline force-app/main/default/classes/DRL_TestUtility.cls:105]:
```suggestion
            objLead.Status = this.strStatus;
```

File: force-app/main/default/classes/DRL_TestUtility.cls [hunk 31]
```diff
@@ -79,7 +101,10 @@ public class DRL_TestUtility {
...
+            objLead.Country=this.strCountry;
```
Comment (csankit0108) [inline force-app/main/default/classes/DRL_TestUtility.cls:106]:
```suggestion
            objLead.Country = this.strCountry;
```

File: force-app/main/default/classes/DRL_TestUtility.cls [hunk 32]
```diff
@@ -79,7 +101,10 @@ public class DRL_TestUtility {
...
+            objLead.Reason_for_Decline__c=this.strReasonForDecline;
```
Comment (csankit0108) [inline force-app/main/default/classes/DRL_TestUtility.cls:107]:
```suggestion
            objLead.Reason_for_Decline__c = this.strReasonForDecline;
```

File: force-app/main/default/classes/LeadTriggerHandler.cls [hunk 33]
```diff
@@ -42,7 +42,22 @@ public with sharing class LeadTriggerHandler extends TriggerHandler {
 
     public override void afterUpdate(){
         checkIfLeadQualifiedFor24HrsLoop();
-        checkForMailerConfigsAfterLeadConversion();  
+        checkForMailerConfigsAfterLeadConversion(); 
+        Set<String> set_oldProspectStatus = new Set<String>();
```
Comment (csankit0108) [inline force-app/main/default/classes/LeadTriggerHandler.cls:46]:
```suggestion
        Set<String> set_OldProspectStatuses = new Set<String>();
```

File: force-app/main/default/classes/LeadTriggerHandler.cls [hunk 34]
```diff
@@ -42,7 +42,22 @@ public with sharing class LeadTriggerHandler extends TriggerHandler {
...
+        Set<String> set_newLeadStatus = new Set<String>();
```
Comment (csankit0108) [inline force-app/main/default/classes/LeadTriggerHandler.cls:47]:
```suggestion
        Set<String> set_NewLeadStatuses = new Set<String>();
```

**Review by csankit0108 — CHANGES_REQUESTED** [review]

File: force-app/main/default/classes/DRL_LeadTriggerHelper.cls [hunk 35]
```diff
@@ -246,5 +246,89 @@ public without sharing class DRL_LeadTriggerHelper {
...
+    * @ author       : Wafa Shaikh
+    * @ description  : Updating the prosepct record with the Status value based on the Lead records status value
+    * @ params       : list_Leads - List of Lead records
+    * @ return       : N/A
+    */
+    public static void processAfterUpdate(List<Lead> list_Leads) {
+        //Set to store the Prospect Ids
+        Set<Id> set_ProspectIds = new Set<Id>();
+        //map to get the weightage of Status value  
+        Map<String, DRL_ProspectLeadStatusMapping__mdt> map_StatusMappingByLeadStatus = new Map<String, DRL_ProspectLeadStatusMapping__mdt>();
```
Comment (pradeepchary) [inline force-app/main/default/classes/DRL_LeadTriggerHelper.cls:261]: break in to two lines

File: force-app/main/default/classes/DRL_LeadTriggerHelper.cls [hunk 36]
```diff
@@ -246,5 +246,89 @@ public without sharing class DRL_LeadTriggerHelper {
...
+            Lead objProspect = map_LeadById.containsKey(objLead.DRL_Prospect__c) 
+                ? map_LeadById.get(objLead.DRL_Prospect__c)
+                : new Lead();
+            //New status value of the Lead
+            String strLeadStatus = objLead.Status;
+            //Current status value of the Prospect
+            String strProspectStatus = objProspect?.Status;
+            //Weightage of the new Lead status value
+            Integer intNewStatusWeightage = map_StatusMappingByLeadStatus.containsKey(strLeadStatus)
+                ? (Integer)map_StatusMappingByLeadStatus.get(strLeadStatus)?.Weightage__c
```
Comment (pradeepchary) [inline force-app/main/default/classes/DRL_LeadTriggerHelper.cls:303]:
```suggestion
                ? (Integer)map_StatusMappingByLeadStatus.get(strLeadStatus).Weightage__c
```
You already did a check that the map contains this particular key. the metadata record will the weightage

File: force-app/main/default/classes/DRL_LeadTriggerHelper.cls [hunk 37]
```diff
@@ -246,5 +246,89 @@ public without sharing class DRL_LeadTriggerHelper {
...
+                : 0;
+            //Weightage of the current status value of the Prospect
+            Integer intOldStatusWeightage = (String.isNotBlank(strProspectStatus) && map_StatusMappingByLeadStatus.containsKey(strProspectStatus))
+                ? (Integer)map_StatusMappingByLeadStatus.get(strProspectStatus)?.Weightage__c
```
Comment (pradeepchary) [inline force-app/main/default/classes/DRL_LeadTriggerHelper.cls:307]:
```suggestion
                ? (Integer)map_StatusMappingByLeadStatus.get(strProspectStatus).Weightage__c
```

File: force-app/main/default/classes/DRL_LeadTriggerHelper.cls [hunk 38]
```diff
@@ -246,5 +246,89 @@ public without sharing class DRL_LeadTriggerHelper {
...
+                : 0;
+            
+            /*Check if the new Lead status weightage > current status of Prospect and
+              old Lead status is not same as the new Prospect status that needs to be updated
+            */
+            if (
+                (intNewStatusWeightage > intOldStatusWeightage) &&
+                map_StatusMappingByLeadStatus.containsKey(strLeadStatus) && 
+                (strProspectStatus != map_StatusMappingByLeadStatus.get(strLeadStatus)?.ProspectStatus__c)
```
Comment (pradeepchary) [inline force-app/main/default/classes/DRL_LeadTriggerHelper.cls:316]:
```suggestion
                (strProspectStatus != map_StatusMappingByLeadStatus.get(strLeadStatus).ProspectStatus__c)
```

File: force-app/main/default/classes/DRL_LeadTriggerHelper.cls [hunk 39]
```diff
@@ -246,5 +246,89 @@ public without sharing class DRL_LeadTriggerHelper {
...
+            ) {
+                objProspect.Status = map_StatusMappingByLeadStatus.get(strLeadStatus)?.ProspectStatus__c;
```
Comment (pradeepchary) [inline force-app/main/default/classes/DRL_LeadTriggerHelper.cls:318]:
```suggestion
                objProspect.Status = map_StatusMappingByLeadStatus.get(strLeadStatus).ProspectStatus__c;
```

File: force-app/main/default/classes/DRL_LeadTriggerHelper.cls [hunk 40]
```diff
@@ -246,5 +246,89 @@ public without sharing class DRL_LeadTriggerHelper {
...
-    }     
+    }
+
+    /*
+    * @ author       : Wafa Shaikh
+    * @ description  : Updating the prosepct record with the Status value based on the Lead records status value
+    * @ params       : list_Leads - List of Lead records
+    * @ return       : N/A
+    */
+    public static void processAfterUpdate(List<Lead> list_Leads) {
```
Comment (pradeepchary) [inline force-app/main/default/classes/DRL_LeadTriggerHelper.cls:257]: the Prospect status update logic on the Lead status update can be pushed to Async method

File: force-app/main/default/classes/DRL_TestUtility.cls [hunk 41]
```diff
@@ -31,7 +35,9 @@ public class DRL_TestUtility {
             this.strCompany = '';
             this.strUTMCampaign = '';
             this.strUTMSource = '';
-            
+            this.strStatus='';
```
Comment (pradeepchary) [inline force-app/main/default/classes/DRL_TestUtility.cls:38]:
```suggestion
            this.strStatus = '';
```

File: force-app/main/default/classes/DRL_TestUtility.cls [hunk 42]
```diff
@@ -31,7 +35,9 @@ public class DRL_TestUtility {
...
+            this.strCountry='';
```
Comment (pradeepchary) [inline force-app/main/default/classes/DRL_TestUtility.cls:39]:
```suggestion
            this.strCountry = '';
```

File: force-app/main/default/classes/DRL_TestUtility.cls (hunk 28 above)
Comment (pradeepchary) [inline force-app/main/default/classes/DRL_TestUtility.cls:91]:
```suggestion
         public LeadBuilder setReasonForDecline(String strValue) {
```

**Review by pradeepchary — APPROVED** [review]

**Comment by pradeepchary** [comment]
Approving this PR, added a few comments, address them, and do test thoroughly the logic and then merge the PR

**Review by csankit0108 — APPROVED** [review]

**Pull request merged**